module/zfs/spa_errlog.c

```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+	    KM_SLEEP);
+
+	int aff_snap_count = 0;
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	uint64_t zap_clone = dsl_dir_phys(ds->ds_dir)->dd_clones;
+
+	/* Check only snapshots created from this file system. */
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg &&
+	    snap_obj_txg <= txg_to_consider) {
+
+		dsl_dataset_rele(ds, FTAG);
+		error = dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds);
+		if (error != 0)
+			goto out;
+
+		if (dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj != fs)
+			break;
+
+		boolean_t affected = B_TRUE;
+		if (check_snapshot) {
+			uint64_t blk_txg;
+			error = find_block_txg(ds, zep, &blk_txg);
+			affected = (error == 0 && zep->zb_birth == blk_txg);
+		}
+
+		if (affected) {
+			snap_obj_array[aff_snap_count] = snap_obj;
+			aff_snap_count++;
+
+			if (!only_count) {
+				zbookmark_phys_t zb;
+				zep_to_zb(snap_obj, zep, &zb);
+				if (copyout(&zb, (char *)addr + (*count - 1) *
+				    sizeof (zbookmark_phys_t),
+				    sizeof (zbookmark_phys_t)) != 0) {
+					dsl_dataset_rele(ds, FTAG);
+					error = SET_ERROR(EFAULT);
+					goto out;
+				}
+				(*count)--;
+			} else {
+				(*count)++;
+			}
+		}
+		snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+		snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	}
+	dsl_dataset_rele(ds, FTAG);
+
+	/* Check clones referencing the error block. */
+	if (zap_clone != 0 && aff_snap_count > 0) {
+		zap_cursor_t zc;
+		zap_attribute_t za;
+
+		for (zap_cursor_init(&zc, spa->spa_meta_objset, zap_clone);
+		    zap_cursor_retrieve(&zc, &za) == 0;
+		    zap_cursor_advance(&zc)) {
+
+			dsl_pool_t *dp = spa->spa_dsl_pool;
+			dsl_dataset_t *clone;
+			error = dsl_dataset_hold_obj(dp, za.za_first_integer,
+			    FTAG, &clone);
+
+			if (error != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+
+			boolean_t found = B_FALSE;
+			for (int i = 0; i < aff_snap_count && !found; i++) {
+				found =
+				    (dsl_dir_phys(clone->ds_dir)->dd_origin_obj
+				    == snap_obj_array[i]);
+			}
+			dsl_dataset_rele(clone, FTAG);
+
+			if (!found)
+				continue;
+
+			error = check_filesystem(spa, za.za_first_integer, zep,
+			    count, addr, only_count);
+
+			if (error != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+		}
+		zap_cursor_fini(&zc);
+	}
+out:
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (error);
+}
+
+static int
+find_top_affected_fs(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
+    uint64_t *top_affected_fs)
+{
+	dsl_dataset_t *ds;
+	dsl_pool_t *dp = spa->spa_dsl_pool;
+
+	int error = dsl_dataset_hold_obj(dp, head_ds, FTAG, &ds);
+	if (error != 0)
+		return (error);
+
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	*top_affected_fs = head_ds;
+
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg) {
```

Comment:
I think that this could walk past this dataset's origin (clone point).  But we might expect that to not happen since zb_birth shouldn't be less than the origin?  Could we add assertions to that effect?

Reply:
Seems reasonable, I was thinking the following: Figure out the `dd_head_dataset_obj` of the current filesystem, get its `ds_creation_txg`, and then assert that it is not greater than the current snapshot's creation txg. Does this make sense?

Reply:
Given your explanation of this function elsewhere, it sounds like the entire point of it involves walking past the origin, to determine which head filesystem contains the first snapshot that contains this block.  So this assertion would be invalid?

Reply:
Yes, you are right! I got confused myself.